internal/scheduler/scheduler.go

```diff
@@ -2,344 +2,660 @@ package scheduler
+	// Am I leader? If not then we're done
+	if !s.leaderController.ValidateToken(leaderToken) {
+		return nil
 	}
-	log.Infof("got updated jobs: %+v", jobs)
-	for _, job := range jobs {
-		job := job
 
-		// Keep the serial number updated.
-		if job.Serial > srv.JobsSerial {
-			srv.JobsSerial = job.Serial
-		}
+	// Start a transaction- we'll roll this back if we don't publish
+	txn := s.jobDb.WriteTxn()
+	defer txn.Abort()
 
-		// If the job is inactive, remove it from the set of active jobs
-		// and remove all runs associated with the job.
-		if job.Cancelled || job.Succeeded || job.Failed {
-			delete(srv.ActiveJobs, job.JobID)
-			delete(srv.RunsByJobId, job.JobID)
-			continue
+	// If we've been asked to generate messages for all jobs do so, else generate messages only for jobs updated this
+	// cycle
+	if updateAll {
+		updatedJobs, err = s.jobDb.GetAll(txn)
+		if err != nil {
+			return err
 		}
+	}
 
-		// If we've never seen this job before, it should be added to the queue.
-		if _, ok := srv.ActiveJobs[job.JobID]; !ok {
-			srv.QueuedJobIds = append(srv.QueuedJobIds, job.JobID)
-		}
+	// Generate any events that came out of synchronising the db state
+	events, err := s.generateUpdateMessages(ctx, updatedJobs)
+	if err != nil {
+		return err
+	}
+
+	// Remove any jobs that have moved into a terminal state as a result of the above actions
+	err = s.removeTerminalJobs(txn, updatedJobs)
+	if err != nil {
+		return err
+	}
 
-		// Add the job to the set of active jobs.
-		srv.ActiveJobs[job.JobID] = &job
+	// Expire any jobs running on clusters that haven't heartbeated within our time limit
+	expirationEvents, err := s.expireJobsIfNecessary(txn)
+	if err != nil {
+		return err
+	}
+	events = append(events, expirationEvents...)
+
+	// Schedule Jobs
+	scheduledJobs, err := s.schedulingAlgo.Schedule(txn, s.jobDb)
+	if err != nil {
+		return err
 	}
+	scheduledJobEvents, err := s.generateLeaseMessages(scheduledJobs)
+	if err != nil {
+		return err
+	}
+	events = append(events, scheduledJobEvents...)
 
-	// New runs for active jobs.
-	i := 0
-	jobIds := make([]uuid.UUID, len(srv.ActiveJobs))
-	for jobId := range srv.ActiveJobs {
-		jobIds[i] = jobId
-		i++
+	// Publish to pulsar
+	err = s.publisher.PublishMessages(ctx, events, leaderToken)
+	if err != nil {
+		return err
 	}
-	runs, err := queries.SelectNewRunsForJobs(ctx, schedulerdb.SelectNewRunsForJobsParams{
-		JobIds: jobIds,
-		Serial: srv.RunsSerial,
-	})
+	txn.Commit()
+	return nil
+}
+
+// syncState updates the state of the jobs in jobDb to match the state in postgres
+// It returns all jobs that have been updated
+func (s *Scheduler) syncState(ctx context.Context) ([]*SchedulerJob, error) {
+	updatedJobs, updatedRuns, err := s.jobRepository.FetchJobUpdates(ctx, s.jobsSerial, s.runsSerial)
 	if err != nil {
-		return errors.WithStack(err)
+		return nil, err
 	}
+	log.Infof("Received %d updated jobs and %d updated job runs", len(updatedJobs), len(updatedRuns))
```

Comment:
Nit: breaking out updateJobs and updatedRuns into logrus fields would enable slightly better log structure/analysis.